imap_processing/cdfutils/global_base.py

```diff
@@ -0,0 +1,353 @@
+    descriptor: str
+    text: str
+
+    def output(self):
+        """
+        Generate the output for the instrument as a dictionary.
+
+        Returns
+        -------
+        dictionary of correctly formatted values for the data_version, descriptor, text,
+         and logical_file_id, added on to the global attributes from GlobalBase
+        """
+        return {
+            "Data_version": self.version,
+            "Descriptor": [self.descriptor],
+            "TEXT": [self.text],
+            "Logical_file_id": ["FILL ME IN AT FILE CREATION"],
+        } | GlobalBase.output()
+
+
+@dataclass
+class DataLevelBase:
+    """
+    Class for all the attributes for the data level base.
+
+    This is used to make the attributes for each data level, and includes
+    InstrumentBase for the required instrument attributes and global attributes.
+
+    Attributes
+    ----------
+    data_type: str
+        The level of data, ex "L1>Level-1"
+    logical_source: str
+        The source of the data, ex "imap_idex_l1"
+    logical_source_desc: str
+        The description of the data, ex "IMAP Mission IDEX Instrument Level-1 Data."
+    instrument_base: InstrumentBase
+        The InstrumentBase object describing the basic instrument information
+    """
+
+    data_type: str
+    logical_source: str
+    logical_source_desc: str
+    instrument_base: InstrumentBase
+
+    def output(self):
+        """
+        Generate the output for the data level as a dictionary.
+
+        Returns
+        -------
+        dictionary of correctly formatted values for the attributes in the class and
+        the attributes from InstrumentBase
+        """
+        return {
+            "Data_type": self.data_type,
+            "Logical_source": self.logical_source,
+            "Logical_source_description": self.logical_source_desc,
+        } | self.instrument_base.output()
+
+
+@dataclass
+class TypeBase:
+    """
+    The general class for attributes, with some reasonable defaults.
+
+    Attributes
+    ----------
+    validmin: np.float64 | np.int64
+        The valid minimum value, required
+    validmax: np.float64 | np.int64
+        The valid maximum value, required
+    display_type: str default=None
+        The display type of the plot (ex "no_plot"), required
+    catdesc: str, default=None
+        The category description, "CATDESC" attribute, required
+    fieldname: str, default=None
+        The fieldname, "FIELDNAM" attribute
+    var_type: str, default="support_data"
+        The type of data
+    fill_val: np.int64, default=Constants.INT_FILLVAL
+        The values for filling data
+    scale_type: str, default="linear"
+        The scale of the axis, "SCALETYP" attribute
+    label_axis: str, default=None
+        Axis label, "LABLAXIS" attribute
+    format: str, default=None
+        The format of the data, in Fortran format
+    units: str, default=None
+        The units of the data
+    """
+
+    validmin: np.float64 | np.int64
```

Comment:
```
    validmin: np.float64 | np.int64
E   TypeError: unsupported operand type(s) for |: 'type' and 'type'
```
I think that might only be for 3.10+, so you might need to do a typing.Union on it? or you could do a `numbers.Real` as well?